configserver/src/main/java/com/yahoo/vespa/config/server/ApplicationRepository.java

```diff
@@ -440,21 +446,21 @@ private RemoteSession getRemoteSession(Tenant tenant, long sessionId) {
         return currentActiveApplicationSet;
     }
 
-    private File decompressApplication(InputStream in, String contentType) {
+    private File decompressApplication(InputStream in, String contentType, File tempDir) {
         try (CompressedApplicationInputStream application =
                      CompressedApplicationInputStream.createFromCompressedStream(in, contentType)) {
-            return decompressApplication(application);
+            return decompressApplication(application, tempDir);
         } catch (IOException e) {
+            cleanupTempDirectory(tempDir);
```

Comment:
Shouldn't this be in a `finally` block, in case a different exception is thrown? Same below.

Reply:
No, then it will always delete it, it shouldn't delete it unless it's an exception (or until it does need the temp dir anymore). But this is a bit confusing and was what led to the bug in the first place, I'll try to rewrite the code to make it clearer.